### PR TITLE
New Item: Golden Arrow

### PR DESCRIPTION
**Pull request by shimaosu**

### CODE REVIEW

File: game/scripts/vscripts/items/item_golden_arrow.lua [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+
+function modifier_item_golden_arrow:GetModifierPhysicalArmorBonus()
+	return self:GetAbility():GetSpecialValueFor("bonus_armor")
+end
+
+if IsServer() then
+	function item_golden_arrow:OnSpellStart()
+		local caster = self:GetCaster()
+		local stacks = caster:GetModifierStackCount("modifier_item_golden_arrow", self)
+		local coof = stacks*self:GetSpecialValueFor("gold_per_stack")
```
Comment (ark120202) [inline game/scripts/vscripts/items/item_golden_arrow.lua]:
```lua

File: game/scripts/vscripts/items/item_golden_arrow.lua [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+end
+
+function modifier_item_golden_arrow:GetModifierPhysicalArmorBonus()
+	return self:GetAbility():GetSpecialValueFor("bonus_armor")
+end
+
+if IsServer() then
+	function item_golden_arrow:OnSpellStart()
+		local caster = self:GetCaster()
+		local stacks = caster:GetModifierStackCount("modifier_item_golden_arrow", self)
```
Comment (ark120202) [inline game/scripts/vscripts/items/item_golden_arrow.lua]:
```lua
local modifier = caster:FindModifierByName("modifier_item_golden_arrow")
if not modifier then
	modifier = caster:AddNewModifier(...)
else
	-- Don't remember if new modifiers already have 1 stack, if they don't, then put it out of else branch
	modifier:IncrementStackCount()
end

local stacks = modifier:GetStackCount()
```

File: game/scripts/vscripts/items/item_golden_arrow.lua [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+		local coof = stacks*self:GetSpecialValueFor("gold_per_stack")
+		local base_gold = (self:GetSpecialValueFor("base_gold")+(caster:GetLevel()*30))
+		local stats_debuff_per_stack = stacks*self:GetSpecialValueFor("stats_debuff_per_stack")
+		local damage = (caster:GetModifierStackCount("modifier_item_golden_arrow", self)) * (self:GetSpecialValueFor("damage_per_stack"))
+		local cooldown = self:GetCooldown(self:GetLevel() ) + ( stacks * (self:GetSpecialValueFor("cooldown_per_stack")))
+
+		caster:EmitSound("Arena.Items.GoldenArrow.Activate")
+
+		ModifyStacksLua(self, caster,caster, "modifier_item_golden_arrow", 1)
```
Comment (ark120202) [inline game/scripts/vscripts/items/item_golden_arrow.lua]: See above

File: game/scripts/vscripts/items/item_golden_arrow.lua [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+end
+
+if IsServer() then
+	function item_golden_arrow:OnSpellStart()
+		local caster = self:GetCaster()
+		local stacks = caster:GetModifierStackCount("modifier_item_golden_arrow", self)
+		local coof = stacks*self:GetSpecialValueFor("gold_per_stack")
+		local base_gold = (self:GetSpecialValueFor("base_gold")+(caster:GetLevel()*30))
+		local stats_debuff_per_stack = stacks*self:GetSpecialValueFor("stats_debuff_per_stack")
+		local damage = (caster:GetModifierStackCount("modifier_item_golden_arrow", self)) * (self:GetSpecialValueFor("damage_per_stack"))
```
Comment (ark120202) [inline game/scripts/vscripts/items/item_golden_arrow.lua]: You can use `stacks` variable there

File: game/scripts/vscripts/items/item_golden_arrow.lua [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+function modifier_item_golden_arrow:GetModifierPhysicalArmorBonus()
+	return self:GetAbility():GetSpecialValueFor("bonus_armor")
+end
+
+if IsServer() then
+	function item_golden_arrow:OnSpellStart()
+		local caster = self:GetCaster()
+		local stacks = caster:GetModifierStackCount("modifier_item_golden_arrow", self)
+		local coof = stacks*self:GetSpecialValueFor("gold_per_stack")
+		local base_gold = (self:GetSpecialValueFor("base_gold")+(caster:GetLevel()*30))
```
Comment (ark120202) [inline game/scripts/vscripts/items/item_golden_arrow.lua]: Remove braces around sole expression
Also, add spaces around binary operators

File: game/scripts/vscripts/items/item_golden_arrow.lua [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+		local stats_debuff_per_stack = stacks*self:GetSpecialValueFor("stats_debuff_per_stack")
+		local damage = (caster:GetModifierStackCount("modifier_item_golden_arrow", self)) * (self:GetSpecialValueFor("damage_per_stack"))
+		local cooldown = self:GetCooldown(self:GetLevel() ) + ( stacks * (self:GetSpecialValueFor("cooldown_per_stack")))
```
Comment (ark120202) [inline game/scripts/vscripts/items/item_golden_arrow.lua]: Have you tried to make it visible in hud by overriding `GetCooldown`? It works in abilities, but not sure about items

File: game/scripts/vscripts/items/item_golden_arrow.lua [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+
+		caster:EmitSound("Arena.Items.GoldenArrow.Activate")
+
+		ModifyStacksLua(self, caster,caster, "modifier_item_golden_arrow", 1)
+
+		ParticleManager:SetParticleControlEnt(ParticleManager:CreateParticle("particles/arena/items_fx/coffee_bean_refresh.vpcf", PATTACH_ABSORIGIN_FOLLOW, caster), 0, caster, PATTACH_POINT_FOLLOW, "attach_hitloc", caster:GetAbsOrigin(), true)
```
Comment (ark120202) [inline game/scripts/vscripts/items/item_golden_arrow.lua]: That line's too long, break it

File: game/scripts/vscripts/items/item_golden_arrow.lua (hunk 5 above)
Comment (ark120202) [inline game/scripts/vscripts/items/item_golden_arrow.lua]: What's `caster:GetLevel()*30`? Don't see that thing in item's description. Anyway, don't use magic numbers in code, put them to specials.

File: game/scripts/vscripts/items/item_golden_arrow.lua [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+
+		if caster:IsRealHero() then
+			if caster:GetStrength() and caster:GetAgility() and caster:GetIntellect() > 5 then
+				caster:ModifyStrength(stats_debuff_per_stack)
+				caster:ModifyAgility(stats_debuff_per_stack)
+				caster:ModifyIntellect(stats_debuff_per_stack)
+
+				Gold:ModifyGold(caster, base_gold)
```
Comment (ark120202) [inline game/scripts/vscripts/items/item_golden_arrow.lua]: Why you don't notify about it with message?

File: game/scripts/vscripts/items/item_golden_arrow.lua (hunk 1 above)
Comment (ark120202) [inline game/scripts/vscripts/items/item_golden_arrow.lua]: `coof` is somewhat cryptic name

File: game/scripts/vscripts/items/item_golden_arrow.lua [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+				Gold:AddGoldWithMessage(caster,coof)
+
+			elseif not caster:HasModifier("modifier_fountain_aura_arena") then
+				ApplyDamage({victim = caster, attacker = caster, damage = damage, damage_type = DAMAGE_TYPE_PURE, ability = self})
```
Comment (ark120202) [inline game/scripts/vscripts/items/item_golden_arrow.lua]: Break that object in multiple lines

File: game/scripts/vscripts/items/item_golden_arrow.lua (hunk 8 above)
Comment (shimaosu) [reply game/scripts/vscripts/items/item_golden_arrow.lua]: cause it is gold that caster loses when dies

File: game/scripts/vscripts/items/item_golden_arrow.lua (hunk 5 above)
Comment (shimaosu) [reply game/scripts/vscripts/items/item_golden_arrow.lua]: it is gold that caster loses when die

File: game/scripts/vscripts/items/item_golden_arrow.lua [hunk 10]
```diff
@@ -95,3 +95,7 @@ modifier_item_golden_arrow_counter = class({
 	IsPurgable = function() return false end,
 	RemoveOnDeath = function() return false end,
 })
+
+function modifier_item_golden_arrow_counter:GetTextureName()
```
Comment (shimaosu) [inline game/scripts/vscripts/items/item_golden_arrow.lua]: it made to avoid modifier texture destroy when you sell item

**Pull request merged**